src/controller/event/handler/internal/artifact.go

```diff
@@ -55,43 +99,137 @@ func (a *Handler) IsStateful() bool {
 }
 
 func (a *Handler) onPull(ctx context.Context, event *event.ArtifactEvent) error {
+	// if duration is equal to 0 or negative, keep original sync mode.
+	if asyncFlushDuration <= 0 {
+		var tagName string
+		if len(event.Tags) > 0 {
+			tagName = event.Tags[0]
+		}
+
+		if !config.PullTimeUpdateDisable(ctx) {
+			a.syncFlushPullTime(ctx, event.Artifact.ID, tagName, time.Now())
+		}
+
+		if !config.PullCountUpdateDisable(ctx) {
+			a.syncFlushPullCount(ctx, event.Artifact.RepositoryID, 1)
+		}
+
+		return nil
+	}
+
+	// async mode, update in cache firstly and flush to db by workers periodically.
+	a.once.Do(func() {
+		if !config.PullTimeUpdateDisable(ctx) {
+			a.pullTimeStore = make(map[string]time.Time)
+			go a.asyncFlushPullTime(orm.Context())
+		}
+
+		if !config.PullCountUpdateDisable(ctx) {
+			a.pullCountStore = make(map[int64]uint64)
+			go a.asyncFlushPullCount(orm.Context())
+		}
+	})
+
 	if !config.PullTimeUpdateDisable(ctx) {
-		go func() { a.updatePullTime(ctx, event) }()
+		a.updatePullTimeInCache(ctx, event)
 	}
+
 	if !config.PullCountUpdateDisable(ctx) {
-		go func() { a.addPullCount(ctx, event) }()
+		a.addPullCountInCache(ctx, event)
 	}
+
 	return nil
 }
 
-func (a *Handler) updatePullTime(ctx context.Context, event *event.ArtifactEvent) {
-	var tagID int64
+func (a *Handler) updatePullTimeInCache(ctx context.Context, event *event.ArtifactEvent) {
+	var tagName string
 	if len(event.Tags) != 0 {
+		tagName = event.Tags[0]
+	}
+
+	key := fmt.Sprintf("%d:%s", event.Artifact.ID, tagName)
+
+	a.pullTimeLock.Lock()
+	defer a.pullTimeLock.Unlock()
+
+	a.pullTimeStore[key] = time.Now()
+}
+
+func (a *Handler) addPullCountInCache(ctx context.Context, event *event.ArtifactEvent) {
+	a.pullCountLock.Lock()
+	defer a.pullCountLock.Unlock()
+
+	a.pullCountStore[event.Artifact.RepositoryID] = a.pullCountStore[event.Artifact.RepositoryID] + 1
+}
+
+func (a *Handler) syncFlushPullTime(ctx context.Context, artifactID int64, tagName string, time time.Time) {
+	var tagID int64
+
+	if tagName != "" {
 		tags, err := tag.Ctl.List(ctx, &q.Query{
 			Keywords: map[string]interface{}{
-				"ArtifactID": event.Artifact.ID,
-				"Name":       event.Tags[0],
+				"ArtifactID": artifactID,
+				"Name":       tagName,
 			},
 		}, nil)
 		if err != nil {
-			log.Infof("failed to list tags when to update pull time, %v", err)
+			log.Warningf("failed to list tags when to update pull time, %v", err)
 		} else {
 			if len(tags) != 0 {
 				tagID = tags[0].ID
 			}
 		}
 	}
-	if err := artifact.Ctl.UpdatePullTime(ctx, event.Artifact.ID, tagID, time.Now()); err != nil {
-		log.Debugf("failed to update pull time form artifact %d, %v", event.Artifact.ID, err)
+
+	if err := artifact.Ctl.UpdatePullTime(ctx, artifactID, tagID, time); err != nil {
```

Comment:
It seems that artifact.Ctl.UpdatePullTime() doesn't handle well the situation of tagID equals 0, please update it to handle the ErrorNotFound.

Reply:
seems we should handle that in the artifact controller?

Reply:
Let's use another pr to fix the situation.